In [2]:
import pandas as pd
from typing import Dict

In [3]:
def analyze_metrics(csv_path: str) -> Dict[str, Dict[str, float]]:
    # Load CSV
    df = pd.read_csv(csv_path)

    # Define metric columns to analyze
    metric_columns = [
        "sequence", "jaccard", "levenshtein", "structure",
        "content", "content2", "f1", "similarity"
    ]

    # 1. Overall averages
    overall_averages = df[metric_columns].mean().round(4).to_dict()

    # 2. Language-level averages
    language_averages = (
        df.groupby("Language")[metric_columns]
        .mean()
        .round(4)
        .to_dict(orient="index")
    )

    return {
        "overall": overall_averages,
        "by_language": language_averages
    }

In [4]:
import pandas as pd
from IPython.display import display

def display_stats(stats: dict, name: str = "Tool"):
    print(f"Evaluation Summary for: {name}\n")

    # Overall average metrics
    overall_df = pd.DataFrame.from_dict(stats["overall"], orient="index", columns=["Overall Average"])
    display(overall_df)

    # Language-level average metrics
    language_df = pd.DataFrame.from_dict(stats["by_language"], orient="index")
    language_df.index.name = "Language"
    display(language_df)


In [4]:
docling_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_docling.csv")
display_stats(docling_stat, "Docling")

Evaluation Summary for: Docling



,Overall Average
sequence,0.2193
jaccard,0.6047
levenshtein,0.3012
structure,0.0652
content,0.8974
content2,0.9100
f1,0.8197
similarity,0.5454


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.1374,0.3519,0.2628,0.0689,0.8370,0.8605,0.4914,0.4300
CS,0.2199,0.5915,0.2849,0.0605,0.9162,0.9217,0.8123,0.5439
DA,0.2406,0.6361,0.3097,0.0699,0.9453,0.9585,0.8685,0.5755
DE,0.2591,0.6500,0.3253,0.0663,0.9478,0.9641,0.8745,0.5839
EL,0.1251,0.3372,0.2085,0.0659,0.4527,0.4817,0.4581,0.3042
EN,0.2387,0.6246,0.3067,0.0643,0.9547,0.9805,0.8729,0.5775
ES,0.2358,0.6425,0.3228,0.0642,0.9583,0.9845,0.8722,0.5829
ET,0.1844,0.6726,0.2933,0.0646,0.9308,0.9247,0.8813,0.5645
FI,0.1761,0.6746,0.3155,0.0622,0.9284,0.9218,0.8783,0.5653


In [5]:
pymupdf_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_unstructures.csv")
display_stats(pymupdf_stat, "unstructured")

Evaluation Summary for: unstructured



,Overall Average
sequence,0.2169
jaccard,0.7030
levenshtein,0.3136
structure,0.0657
content,0.9474
content2,0.9389
f1,0.9160
similarity,0.5859


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.1978,0.6639,0.3158,0.0639,0.9721,0.9663,0.8886,0.5812
CS,0.2017,0.7424,0.2938,0.0628,0.9570,0.9460,0.9402,0.5920
DA,0.2159,0.7021,0.3140,0.0669,0.9728,0.9636,0.9284,0.5948
DE,0.2743,0.7138,0.3303,0.0652,0.9738,0.9656,0.9300,0.6076
EL,0.1668,0.4297,0.3148,0.0660,0.8600,0.8551,0.6577,0.4786
EN,0.2444,0.6969,0.3141,0.0662,0.9817,0.9771,0.9384,0.6027
ES,0.2275,0.7139,0.3300,0.0644,0.9832,0.9802,0.9350,0.6049
ET,0.2013,0.7349,0.2975,0.0662,0.9391,0.9232,0.9361,0.5855
FI,0.2037,0.7379,0.3182,0.0648,0.9386,0.9232,0.9347,0.5887


In [6]:
pymupdf_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_pdfplumber.csv")
display_stats(pymupdf_stat, "pdfplumber")

Evaluation Summary for: pdfplumber



,Overall Average
sequence,0.1222
jaccard,0.4281
levenshtein,0.2723
structure,0.0142
content,0.9265
content2,0.9521
f1,0.6434
similarity,0.4798


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.0763,0.4119,0.2634,0.0132,0.9449,0.9715,0.6263,0.4725
CS,0.1251,0.4445,0.2668,0.0141,0.9157,0.9412,0.6594,0.4810
DA,0.1316,0.4171,0.2769,0.0153,0.9362,0.9611,0.6373,0.4822
DE,0.1243,0.4188,0.2768,0.0144,0.9335,0.9589,0.6370,0.4805
EL,0.0984,0.4140,0.2587,0.0135,0.9306,0.9569,0.6270,0.4713
EN,0.1271,0.3964,0.2773,0.0132,0.9467,0.9729,0.6163,0.4785
ES,0.1193,0.4107,0.2788,0.0132,0.9535,0.9787,0.6257,0.4828
ET,0.1140,0.4541,0.2699,0.0158,0.8931,0.9150,0.6710,0.4761
FI,0.1240,0.4499,0.2707,0.0165,0.8958,0.9182,0.6628,0.4768


In [5]:
pymupdf_stat = analyze_metrics("/home/4baba/EUR_lex/evaluation_results/test_evaluation_pymupdf_hybrid.csv")
display_stats(pymupdf_stat, "pymupdf_hybrid")

Evaluation Summary for: pymupdf_hybrid



,Overall Average
sequence,0.1143
jaccard,0.4288
levenshtein,0.2728
structure,0.0146
content,0.9269
content2,0.9588
f1,0.6442
similarity,0.4801


,sequence,jaccard,levenshtein,structure,content,content2,f1,similarity
Language,,,,,,,,
BG,0.0761,0.4123,0.2656,0.0137,0.9406,0.9747,0.6268,0.4728
CS,0.1214,0.4454,0.2666,0.0145,0.9180,0.9496,0.6604,0.4823
DA,0.1185,0.4177,0.2779,0.0156,0.9348,0.9667,0.6382,0.4814
DE,0.1271,0.4198,0.2771,0.0149,0.9316,0.9641,0.6381,0.4818
EL,0.0954,0.4146,0.2602,0.0139,0.9281,0.9613,0.6277,0.4716
EN,0.1191,0.3971,0.2765,0.0136,0.9413,0.9753,0.6173,0.4772
ES,0.1058,0.4113,0.2792,0.0136,0.9478,0.9805,0.6263,0.4806
ET,0.1020,0.4548,0.2708,0.0161,0.9032,0.9304,0.6717,0.4784
FI,0.1039,0.4506,0.2739,0.0168,0.9034,0.9316,0.6636,0.4777


### Combined Evaluation Summary

| Metric        | Docling | Unstructured | pdfplumber | pymupdf_hybrid |
|---------------|--------:|-------------:|-----------:| --------------:
| sequence      | 0.2193  | 0.2169       | 0.1222     | 0.1143
| jaccard       | 0.6047  | 0.7030       | 0.4281     | 0.4288
| levenshtein   | 0.3012  | 0.3136       | 0.2723     | 0.2728
| structure     | 0.0652  | 0.0657       | 0.0142     | 0.0146
| content       | 0.8974  | 0.9474       | 0.9265     | 0.9269
| content2      | 0.9100  | 0.9389       | 0.9521     | 0.9588
| f1            | 0.8197  | 0.9160       | 0.6434     | 0.6442


Checking for '/home/4baba/EUR_lex/pdfs_2024/category10/law32024D3127/32024D3127_EN.pdf'

docling:
<table><tbody><tr><td></td><th colspan="3">Volume of euro coins approved for issuance in 2025</th></tr><tr><td></td><th>Circulation coins</th><th>Collector coins (not intended for circulation)</th><th>Volume of coin issuance</th></tr><tr><td></td><td></td><th>(EUR million)</th><th>(EUR million)</th></tr><tr><td></td><th>(EUR million)</th><td></td><td></td></tr><tr><td>Belgium</td><td>30,00</td><td>0,30</td><td>30,30</td></tr><tr><td>Germany</td><td>343,50</td><td>135,50</td><td>479,00</td></tr><tr><td>Estonia</td><td>11,20</td><td>0,34</td><td>11,54</td></tr><tr><td>Ireland</td><td>30,78</td><td>0,50</td><td>31,28</td></tr><tr><td>Greece</td><td>152,20</td><td>0,96</td><td>153,16</td></tr><tr><td>Spain</td><td>294,94</td><td>40,00</td><td>334,94</td></tr><tr><td>France</td><td>276,47</td><td>60,00</td><td>336,47</td></tr><tr><td>Croatia</td><td>71,92</td><td>0,45</td><td>72,37</td></tr><tr><td>Italy</td><td>241,65</td><td>5,05</td><td>246,70</td></tr><tr><td>Cyprus</td><td>6,30</td><td>0,01</td><td>6,31</td></tr><tr><td>Latvia</td><td>5,90</td><td>0,20</td><td>6,10</td></tr><tr><td>Lithuania</td><td>12,00</td><td>0,63</td><td>12,63</td></tr><tr><td>Luxembourg</td><td>13,00</td><td>0,40</td><td>13,40</td></tr><tr><td>Malta</td><td>8,20</td><td>11,50</td><td>19,70</td></tr><tr><td>Netherlands</td><td>24,00</td><td>1,00</td><td>25,00</td></tr><tr><td>Austria</td><td>114,21</td><td>175,51</td><td>289,72</td></tr><tr><td>Portugal</td><td>35,10</td><td>1,70</td><td>36,80</td></tr><tr><td>Slovenia</td><td>22,00</td><td>1,00</td><td>23,00</td></tr><tr><td>Slovakia</td><td>20,00</td><td>3,00</td><td>23,00</td></tr><tr><td>Finland</td><td>10,00</td><td>5,00</td><td>15,00</td></tr><tr><td>Total</td><td>1 723,37</td><td>443,05</td><td>2 166,42</td></tr></tbody></table>

pymupdf and +added tables/:
Use page.get_text("html") → includes unstyled visual text blocks
Tables are added after text using find_tables(...)

<h2>Page 2</h2>
<div id="page0" style="width:595.3pt;height:841.9pt">
<p style="top:244.7pt;left:85.0pt;line-height:9.6pt"><span style="font-family:EUAlbertina,serif;font-size:9.6pt;color:#000000">Belgium</span></p>
<p style=""><span style="">30,00</span></p>
<p style=""><span style="">0,30</span></p>
<p style=""><span style="">30,30</span></p>
<p style=""><span style="">Germany</span></p>
<p style=""><span style="">343,50</span></p>
<p style=""><span style="">135,50</span></p>
<p style=""><span style="">479,00</span></p>
<p style=""><span style="">Estonia</span></p>
<p style=""><span style="">11,20</span></p>
<p style=""><span style="">0,34</span></p>
<p style=""><span style="">11,54</span></p>
</div>

pymupdf_hybrid:
Parses both:
Text blocks (page.get_text("dict"))
Tables (find_tables(...))
Stores each with a vertical Y position (bbox[1])
Then sorts + merges text and tables by position on the page

both this and the last one use find_table method but they are different in rendering the results

<h3>Table 1</h3>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>App</td>
      <td>roval of the</td>
      <td>volume of euro</td>
      <td>coins to be issued in 2025</td>
      <td></td>
    </tr>
    <tr>
      <td>ECB hereby approves the vol</td>
      <td>ume of euro</td>
      <td>coins to be issu</td>
      <td>ed by euro area Member States in</td>
      <td>2025</td>
    </tr>
    <tr>
      <td>wing table:</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td>Volume of euro c</td>
      <td>oins approved for issuance in 2025</td>
      <td></td>
    </tr>
    <tr>
      <td>Circ</td>
      <td>ulation coins</td>
      <td>Col</td>
      <td>lector coins Vol</td>
      <td>ume of c</td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td>(not inten</td>
      <td>ded for circulation)</td>
      <td>issuance</td>
    </tr>
    <tr>
      <td>(E</td>
      <td>UR million)</td>
      <td>(E</td>
      <td>UR million) (E</td>
      <td>UR millio</td>
    </tr>
    <tr>
      <td>Belgium</td>
      <td>30,00</td>
      <td></td>
      <td>0,30</td>
      <td>30,30</td>
    </tr>
    <tr>
      <td>Germany</td>
      <td>343,50</td>
      <td></td>
      <td>135,50</td>
      <td>479,00</td>
    </tr>

pymupdf_pdfplumber:

Uses pdfplumber.extract_text() for text
Uses PyMuPDF only for tables (on the result of pdfplumber)
Text and tables are separated, but cleaner

Volume of euro coins approved for issuance in 2025
Circulation coins Collector coins Volume of coin
(not intended for circulation) issuance
(EUR million) (EUR million) (EUR million)
Belgium 30,00 0,30 30,30
Germany 343,50 135,50 479,00
Estonia 11,20 0,34 11,54
Ireland 30,78 0,50 31,28
Greece 152,20 0,96 153,16
Spain 294,94 40,00 334,94
France 276,47 60,00 336,47
Croatia 71,92 0,45 72,37
Italy 241,65 5,05 246,70
Cyprus 6,30 0,01 6,31
Latvia 5,90 0,20 6,10
Lithuania 12,00 0,63 12,63
Luxembourg 13,00 0,40 13,40
Malta 8,20 11,50 19,70
Netherlands 24,00 1,00 25,00
Austria 114,21 175,51 289,72
Portugal 35,10 1,70 36,80
Slovenia 22,00 1,00 23,00
Slovakia 20,00 3,00 23,00
Finland 10,00 5,00 15,00
Total 1 723,37 443,05 2 166,42


unstructures:
<br/>
<table><thead><tr><th></th><th rowspan="2">Circulation coins (EUR million)</th><th rowspan="2">Collector coins (not intended for circulation) (EUR million)</th><th rowspan="2">Volume of coin issuance (EUR million)</th></tr><tr><th></th></tr></thead><tbody><tr><td>Belgium</td><td>30,00</td><td>0,30</td><td>30,30</td></tr><tr><td>Germany</td><td>343,50</td><td>135,50</td><td>479,00</td></tr><tr><td>Estonia</td><td>11,20</td><td>0,34</td><td>11,54</td></tr><tr><td>Ireland</td><td>30,78</td><td>0,50</td><td>31,28</td></tr><tr><td>Greece</td><td>152,20</td><td>0,96</td><td>153,16</td></tr><tr><td>Spain</td><td>294,94</td><td>40,00</td><td>334,94</td></tr><tr><td>France</td><td>276,47</td><td>60,00</td><td>336,47</td></tr><tr><td>Croatia</td><td>71,92</td><td>0,45</td><td>72,37</td></tr><tr><td>taly</td><td>241,65</td><td>5,05</td><td>246,70</td></tr><tr><td>Cyprus</td><td>6,30</td><td>0,01</td><td>6,31</td></tr><tr><td>Latvia</td><td>5,90</td><td>0,20</td><td>6,10</td></tr><tr><td>Lithuania</td><td>12,00</td><td>0,63</td><td>12,63</td></tr><tr><td>Luxembourg</td><td>13,00</td><td>0,40</td><td>13,40</td></tr><tr><td>Malta</td><td>8,20</td><td>11,50</td><td>19,70</td></tr><tr><td>etherlands</td><td>24,00</td><td>1,00</td><td>25,00</td></tr><tr><td>Austria</td><td>114,21</td><td>175,51</td><td>289,72</td></tr><tr><td>ortugal</td><td>35,10</td><td>1,70</td><td>36,80</td></tr><tr><td>Slovenia</td><td>22,00</td><td>1,00</td><td>23,00</td></tr><tr><td>Slovakia</td><td>20,00</td><td>3,00</td><td>23,00</td></tr><tr><td>Finland</td><td>10,00</td><td>5,00</td><td>15,00</td></tr></tbody></table>
<br/>



The ECB hereby approves the volume of euro coins to be issued by euro area Member States in 2025 as set out in the following table:\n\n| Country   | Circulation coins (EUR million) | Collector coins (not intended for circulation) (EUR million) | Volume of coin issuance (EUR million) |\n
|-----------|---------------------------------|-------------------------------------------------------------|---------------------------------------|\n| Belgium   | 30,00                           | 0,30                                                        | 30,30                                 |\n| Germany   | 343,50                          | 135,50                                                      | 479,00                                |\n| Estonia   | 11,20                           | 0,34                                                        | 11,54                                 |\n| Ireland   | 30,78                           | 0,50                                                        | 31,28                                 |\n| Greece    | 152,20                          | 0,96                                                        | 153,16                                |\n| Spain     | 294,94                          | 40,00                                                       | 334,94                                |\n| France    | 276,47                          | 60,00                                                       | 336,47                                |\n| Croatia   | 71,92                           | 0,45                                                        | 72,37                                 |\n| Italy     | 241,65                          | 5,05                                                        | 246,70                                |\n| Cyprus    | 6,30                            | 0,01                                                        | 6,31                                  |\n| Latvia    | 5,90                            | 0,20                                                        | 6,10                                  |\n| Lithuania | 12,00                           | 0,63                                                        | 12,63                                 |\n| Luxembourg| 13,00                           | 0,40                                                        | 13,40                                 |\n| Malta     | 8,20                            | 11,50                                                       | 19,70                                 |\n| Netherlands| 24,00                          | 1,00                                                        | 25,00                                 |\n| Austria   | 114,21                          | 175,51                                                      | 289,72                                |\n| Portugal  | 35,10                           | 1,70                                                        | 36,80                                 |\n| Slovenia  | 22,00                           | 1,00                                                        | 23,00                                 |\n| Slovakia  | 20,00                           | 3,00                                                        | 23,00                                 |\n| Finland   | 10,00                           | 5,00                                                        | 15,00                                 |\n| **Total** | **1 723,37**                    | **443,05**                                                  | **2 166,42**                         |"